PythonVisualizations/QuadTree.py

```diff
@@ -0,0 +1,227 @@
+            text  = self.canvas.create_text(x-15, y - 12, text = d, fill = "red", font = ("Helvetica", "10"))
+            self.points.append(oval)
+            node.dataObjects = [text]
+      
+            self.COUNTER +=1 #keeps track of number of nodes inserted
+            
+            handler = lambda e: self.setArguments(str(x), str(y), str(d))
+            self.canvas.tag_bind(oval, '<Button>', handler)            
+            
+            return node
+        
+        # if the point to be inserted is identical to the current node, 
+        # add the data to the list of data, but don't recurse any further
+        if n.x == x and n.y == y:
+            n.data.append(d)
+            n.data.append(d)
+            key = d
+            text = self.canvas.create_text((x-3 + n.countSpaces), y - 12, fill = "red",  text = ",", font = ("Helvetica", "10"))
+            n.countSpaces+= 25
+            text  = self.canvas.create_text((x-15 + n.countSpaces), y - 12, fill = "red",  text = key, font = ("Helvetica", "10"))
+            n.dataObjects.append(text)
+            self.COUNTER +=1
+            return n
+      
+
+        # recurse down into the appropriate quadrant
+        self.parent = n
+        if   x >= n.x and y >= n.y:
+            self.direction = SE
+            n.SE = self.__insert(n.SE, x, y, d)
+        elif x >= n.x and y <  n.y:
+            self.direction = NE
+            n.NE = self.__insert(n.NE, x, y, d)
+        elif x <  n.x and y >= n.y: 
+            self.direction = SW
+            n.SW = self.__insert(n.SW, x, y, d)
+        else:
+            self.direction = NW
+            n.NW = self.__insert(n.NW, x, y, d)
+        
+        self.cleanUp(callEnviron)
+        return n    
+    
+    #clears canvas, and deletes all the nodes
+    #with accompanying data and lines
+    def new(self):
+        for i in self.points:
+            self.canvas.delete(i)
+        for i in self.nodes:
+            for j in i.dataObjects:
+                self.canvas.delete(j)
+            i.dataObjects = []
+        for i in self.lines:
+            self.canvas.delete(i)
+        self.canvas.delete(self.rootOutline)
+        self.points = []
+        self.lines= []
+        self.nodes = []
+        self.direction = None
+        self.parent = None
+        self.COUNTER = 0
+        self.__root = None
+
+    #does not allow a data point to be re-used
+    def clickInsert(self):
+        if self.validArgument():
+            x, y, d = self.validArgument()
+            for i in self.nodes:
+                for j in i.data: 
+                    if d== j:
+                        self.setMessage("Node with this data already exists")
+                        return
+            self.insert(x,y,d)
+            msg = "Value {} inserted".format(d)
+        else:
+            msg = "Insert valid argument" 
+        self.setMessage(msg)
```

Comment:
This error message can be a little cryptic for users who don't know that space characters aren't allowed in labels, or the actual dimensions of the canvas.  Perhaps `validArgument()` should return error strings for every *invalid* argument.  If all the strings are None or empty, then all three are valid and can be passed to `insert()`.  Otherwise the messages are concatenated with a separator and shown as the output message.  (Note: embedding newlines in the message will make it span multiple lines).